storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,22 +3582,401 @@ dberr_t fsp_sys_tablespace_validate()
+}
+
+/** Free the extents, fragment page from the given inode
+@param page_no index node page number
+@param offset  index node offset within page
+@return error code */
+static void fseg_inode_free(uint32_t page_no, uint16_t offset)
+{
+  fil_space_t *space= fil_system.sys_space;
+  mtr_t mtr;
+  mtr.start();
+  mtr.x_lock_space(space);
+  buf_block_t *iblock= buf_page_get_gen(page_id_t{0, page_no}, 0,
+                                        RW_X_LATCH, nullptr, BUF_GET,
+                                        &mtr);
+
+  fseg_inode_t *inode= nullptr;
+  DBUG_EXECUTE_IF("unused_undo_free_fail_4", iblock= nullptr;);
+  if (!iblock)
+    goto func_exit;
+
+  sql_print_information("InnoDB: Freeing the unused segment in page "
+                        UINT32PF, page_no);
+  inode= iblock->page.frame + offset;
+  while (fseg_free_step_low(iblock, inode, space, &mtr,
+                            nullptr, true) == DB_SUCCESS_LOCKED_REC)
+  {
+    DBUG_EXECUTE_IF("unused_undo_free_fail_5",
+                    mtr.commit();
+                    return;);
+    iblock->fix();
+    mtr.commit();
+
+    mtr.start();
+    mtr.x_lock_space(space);
+    iblock->page.lock.x_lock();
+    mtr.memo_push(iblock, MTR_MEMO_PAGE_X_FIX);
+  }
+func_exit:
+  mtr.commit();
+}
+
+/** Free the unused segment
+@param boffset    offset of the list containing segment header
+@param space      system tablespace
+@param usedInodes Used system tablespace index node */
+void inode_info::free_segs()
+{
+  for (auto i : *this)
+    fseg_inode_free(uint32_t(i >> 32), uint16_t(i));
+}
+
+/** Remove the unused segment in tablespace. This function
+used only during shrinking of system tablespace
+@param shutdown called during slow shutdown
+@return error code */
+dberr_t fil_space_t::garbage_collect(bool shutdown)
+{
+  if (!shutdown && !trx_sys.is_undo_empty())
+  {
+    sql_print_warning("InnoDB: Cannot free the unused segments"
+                      " in system tablespace because a previous"
+                      " shutdown was not with"
+                      " innodb_fast_shutdown=0");
+    return DB_SUCCESS;
+  }
+
+  ut_a(id == 0);
+  /* Collect all the used segment inode entries */
+  mtr_t mtr;
+  mtr.start();
+  inode_info used_inodes, unused_inodes;
+  dberr_t err= fsp_get_sys_used_segment(&used_inodes, &mtr);
+  DBUG_EXECUTE_IF("unused_undo_free_fail_1", err= DB_CORRUPTION;);
+
+  if (err == DB_SUCCESS)
+    err= used_inodes.get_unused(FSP_HEADER_OFFSET + FSP_SEG_INODES_FULL,
+                                &unused_inodes);
+  DBUG_EXECUTE_IF("unused_undo_free_fail_2", err= DB_CORRUPTION;);
+  if (err == DB_SUCCESS)
+    err= used_inodes.get_unused(FSP_HEADER_OFFSET + FSP_SEG_INODES_FREE,
+                                &unused_inodes);
+  DBUG_EXECUTE_IF("unused_undo_free_fail_3", err= DB_CORRUPTION;);
+  mtr.commit();
+  if (err)
+    return err;
+
+  /* Ignore the error just like trx_purge_free_segment() */
+  unused_inodes.free_segs();
+  return DB_SUCCESS;
```

Comment:
It is up to the caller to ignore the error (after reporting it). This had better be as follows:
```c++
  return unused_inodes.free_segs();
```